keps/sig-node/4815-dra-partitionable-devices/README.md

```diff
@@ -0,0 +1,2137 @@
+	// overwritten by the ones provided here.
+	//
+	// The maximum number of attributes and capacities combined is 32.
+	//
+	// +optional
+	Capacity map[QualifiedName]DeviceCapacity `json:"capacity,omitempty"`
+}
+
+// Device represents one individual hardware instance that can be selected based
+// on its attributes. Besides the name, exactly one field must be set.
+// +k8s:deepcopy-gen=true
+type Device struct {
+	// Name is unique identifier among all devices managed by
+	// the driver in the pool. It must be a DNS label.
+	//
+	// +required
+	Name string `json:"name"`
+
+	...
+
+	// Composite defines one composite device instance.
+	//
+	// +optional
+	// +oneOf=deviceType
+	Composite *CompositeDevice `json:"composite,omitempty"`
+}
+
+// CompositeDevice defines one device instance.
+type CompositeDevice struct {
+	// Includes defines the set of device mixins that this device includes.
+	//
+	// The propertes of each included mixin are applied to this device in
+	// order. Conflicting properties from multiple mixins are taken from the
+	// last mixin listed that contains them.
+	//
+	// The maximum number of mixins that can be included is 8.
+	//
+	// +optional
+	Includes []DeviceMixinRef `json:"includes,omitempty"`
+
+	// ConsumesCapacityFrom defines the set of devices where any capacity
+	// consumed by this device should be pulled from. This applies recursively.
+	// In cases where the dvice names itself as its source, the recursion is
+	// halted.
+	//
+	// Conflicting capacities from multiple devices are taken from the
+	// last device listed that contains them.
+	//
+	// The maximum number of devices that can be referenced is 8.
+	//
+	// +optional
+	ConsumesCapacityFrom []DeviceRef `json:"consumesCapacityFrom,omitempty"`
+
+	// Attributes defines the set of attributes for this device.
+	// The name of each attribute must be unique in that set.
+	//
+	// To ensure this uniqueness, attributes defined by the vendor
+	// must be listed without the driver name as domain prefix in
+	// their name. All others must be listed with their domain prefix.
+	//
+	// Conflicting attributes from those provided via mixins are
+	// overwritten by the ones provided here.
+	//
+	// The maximum number of attributes and capacities combined is 32.
+	//
+	// +optional
+	Attributes map[QualifiedName]DeviceAttribute `json:"attributes,omitempty"`
+
+	// Capacity defines the set of capacities for this device.
+	// The name of each capacity must be unique in that set.
+	//
+	// To ensure this uniqueness, capacities defined by the vendor
+	// must be listed without the driver name as domain prefix in
+	// their name. All others must be listed with their domain prefix.
+	//
+	// Conflicting capacities from those provided via mixins are
+	// overwritten by the ones provided here.
+	//
+	// The maximum number of attributes and capacities combined is 32.
+	//
+	// +optional
+	Capacity map[QualifiedName]DeviceCapacity `json:"capacity,omitempty"`
```

Comment:
Yes, this is a limitation of the current proosal. A sink device *must* consume all of its capacity from some source device - it is unable to define "local" capacities not consumed from elsewhere. As you suggest, a boolean field in the `DeviceCapacity` struct could be added to denote this if we deem it important.

Reply:
Meaning when you declare the capacity you point to the exact source that it should be pulled from? It would make each individual capacity entry a bit longer, but it's definitely more explicit (and readable) in terms of knowing where the capacity is coming from. I'm not opposed to this at all (and actually quite in favor of it).

Reply:
Actually -- thinking about this a bit more, this isn't really possible in practice. We want to be able to define capacities in a mixin, include that mixin in a concrete device, and then (as a property of the device) declare which source it pulls that capacity from (e.g. a specific `mig-1g.5gb` device on `gpu-0` vs. `gpu-1`). If we embed that source information directly in the `Capacity` declaration, we can't define it as a reusable mixin.